### ISSUE TITLE
[Improvement] Rewrite protoc-gen-swiftgrpc with Stencil

### ISSUE THREAD

**Issue by AttilaTheFun** [grpc/grpc-swift#194]
Hey @timburks and @MrMage! I was looking into possibly contributing back to help improve protoc-gen-swiftgrpc to make the generated code a lot prettier, but after poking around a bit I was overwhelmed by all of the formatting code e.g.:
```
    println("/// Main server for generated service")
    println("\(access) final class \(serverName): ServiceServer {")
    indent()
    println("private let provider: \(providerName)")
    println()
    println("\(access) init(address: String, provider: \(providerName)) {")
    ...
```

This is hard to read, repetitive and error prone. I maintain a couple of open source libraries for working with and generating code from OpenAPI 2.0 f.k.a. Swagger documents:
https://github.com/AphelionApps/SwaggerParser
https://github.com/AphelionApps/SwaggerStencil
The SwaggerStencil library combines SwaggerParser and Kyle Fuller's wonderful [Stencil](https://github.com/kylef/Stencil) library to populate elegant templates with type safe SwaggerParser objects.

If you're interested and willing to add the dependency, I'd be happy to help port it over. I've also written  some nice string utilities for turning HTTP paths into pretty(ish) function names which you may find valuable regardless. (It's MIT-licensed)
https://github.com/AphelionApps/SwaggerStencil/blob/master/Sources/String%2BExtensions.swift

**Comment by MrMage** [comment]
Hi Attila! Sorry for the radio silence on this. I do not feel strongly about this, but @timburks [recently changed the template code](https://github.com/grpc/grpc-swift/commit/ae46487c76f860ded3bcbefbbd6685b11d1059d6) from Stencil to the same approach `swift-protobuf` uses, so I suppose it's going to stay that way for now, especially as this also provides consistency with `swift-protobuf`.

**Comment by AttilaTheFun** [comment]
@MrMage what was the motivation behind the new template code? Did you want to remove the dependency or something? It definitely feels like a regression in terms of readability.

**Comment by MrMage** [comment]
I'd mostly refer to @timburks on that one, but I think it was for maintainability and consistency with `swift-protobuf`. For what it's worth, I don't have a clear favorite, given that dealing with all the transformation functions in the old templates could be confusing. Plus, we can more easily re-use some snippets (e.g. streaming send functions) in the templates for several different types of RPCs (server-streaming, bidirectional streaming etc.).

For working with these templates, the best approach I found was to make modifications to the template generator, re-generate `echo.grpc.swift`, then look at the diff. That way I find it easier to parse the actual impact of codegen changes.

**Comment by timburks** [comment]
@AttilaTheFun I originally used Stencil for templates, but had repeated problems with the external dependency:

- Stencil itself adds more dependencies that signficantly increased the build time for the plugin. That was discouraging us from building the plugin in the "main" swift-grpc builds. With the change, the plugin builds very quickly so this is not a concern.

- I had also encountered problems with mirror support in swift-protobuf generated code - at one time (and perhaps now), it was impossible to use Mirrors to enumerate properties of swift-protobuf generated code (they were dynamic). I worked around this by replicating the main data structures of FileDescriptorSet, but this seemed like an ugly solution.

- Also, adding to @MrMage's comments, my own approach for working on the generator has been to rewrite the Echo sample into the form that I wanted and then adjust the generator until it produced that. With this approach, I found it easier to modify the new "printing" code than templates, partly because it's easier to single-step through that code.

My conclusion was that while templates helped with early readability, in my judgement, they detracted from maintainability. I hope you don't find this too much of a barrier.
